### PR TITLE
Integrate OpenMIBOOD datasets and additional benchmark detector with some changes in existing detectors 

### PR DESCRIPTION
**Pull request by taneeshk12**
This PR integrates the OpenMIBOOD medical imaging OOD benchmark into `pytorch-ood`, including the `OpenMIBOODDataset` base class and datasets such as MIDOG, FNAC2019, CHAOS, Task02Heart, OASIS3, BraTS, and ATLAS. It also adds the `Residual` OOD detector and benchmark examples for evaluating existing and OpenMIBOOD-specific methods such as Logit Matching and RankFeat. Additionally, `fDBD` now supports configurable distance normalization, `GradNorm` has a more robust batched prediction fallback.

### CODE REVIEW

File: examples/benchmarks/interface/midog_openmibood.py [hunk 1]
```diff
@@ -7,46 +7,318 @@
 (CVPR 2025).
 
 .. note::
-    The MIDOG data must be prepared first by following the
-    `OpenMIBOOD setup guide <https://github.com/remic-othr/OpenMIBOOD>`_.
-    The ``root`` argument should point at the directory whose subfolders match
-    the relative paths in the bundled image list files (e.g. ``1a/017/...``).
+    Hyperparameters for KNN, DICE, fDBD, GEN, ASH, ReAct, NNGuide, and RankFeat
+    are tuned on the validation set according to the paper's protocol.
+    GradNorm is omitted as it is not part of the paper's 24 benchmarked methods.
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:12]: remove

File: examples/benchmarks/interface/midog_openmibood.py [hunk 2]
```diff
@@ -7,46 +7,318 @@
...
+    DICE,
+    # --- activation pruning ---
+    ASH,
+    ReAct,
+    RankFeat,
+)
 from pytorch_ood.utils import fix_random_seed
 
+# ---------------------------------------------------------------------------
+# Metric Fix: Min-Max Score Scaling
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:52]: I think we should actually address the underlying issue and maybe adjust OOMetrics itself

File: examples/benchmarks/interface/midog_openmibood.py [hunk 3]
```diff
@@ -7,46 +7,318 @@
...
+    file_name="midog_classifier.pth"
+)
+model.load_state_dict(state_dict)
+model = model.eval().to(device)
+
+# Feature extractor (all layers except final FC) — flattened 2048-d, used by most detectors
+feature_extractor = torch.nn.Sequential(*list(model.children())[:-1], torch.nn.Flatten())
+
+# L2-normalized feature extractor (used by KNN, matching OpenMIBOOD paper)
+class L2Norm(torch.nn.Module):
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:94]: we might want to add this as a feature to kNN instead

Comment (kkirchheim) [reply examples/benchmarks/interface/midog_openmibood.py:94]: i guess it depends on weather this is common

File: examples/benchmarks/interface/midog_openmibood.py [hunk 4]
```diff
@@ -7,46 +7,318 @@
...
+    transform=trans,
+    download=True,
+)
+train_loader = torch.utils.data.DataLoader(benchmark.train_set(), batch_size=64)
+
+
+# %%
+# --- Aligned KLMatching Implementation ---
+# OpenMIBOOD's KLMatching differs from standard pytorch-ood in two ways:
+# 1. Reference distributions are conditioned on predicted class (argmax logits), not ground truth.
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:134]: we might want to add this as a feature to KL-matching. also, we should look up if this is actually an error in pytorch-ood.

File: examples/benchmarks/interface/midog_openmibood.py [hunk 5]
```diff
@@ -7,46 +7,318 @@
...
+print("Fitting NNGuide (k=5)...")
+nnguide.fit(train_loader)
+print("Fitting fDBD...")
+fdbd.fit(train_loader)
+print("Fitting DICE (p=75)...")
+dice.fit(train_loader)
+print("Fitting KLMatching...")
+klm.fit(train_loader)
+print("Fitting ReAct (p=95)...")
+react.threshold = 0.5583422333002087
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:249]: where does this hyperparameter come from?

File: examples/benchmarks/interface/midog_openmibood.py [hunk 6]
```diff
@@ -7,46 +7,318 @@
...
+
+print("Fitting ViM...")
+vim.fit(train_loader)
+print("Fitting Residual...")
+residual.fit(train_loader)
+print("Fitting Mahalanobis...")
+mds.fit(train_loader)
+import scipy.linalg
+# --- Mahalanobis Precision Fix ---
+# pytorch-ood's fit injects `1e-6 * I` to the scatter matrix. Since N=1896 < D=2048, the matrix is
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:216]: I think we should actually make this "noise" of 1e-6  configurable in Mahalanobis distance and related detectors. Setting noise to zero should then prevent this problem, right?

File: examples/benchmarks/interface/midog_openmibood.py [hunk 7]
```diff
@@ -7,46 +7,318 @@
...
+klm.fit(train_loader)
+print("Fitting ReAct (p=95)...")
+react.threshold = 0.5583422333002087
 
 # %%
-benchmark = MIDOG_OpenMIBOOD(root="data/openmibood/midog", transform=trans)
+# Standard detectors
+
+from pytorch_ood.api import Detector
+class OpenMIBOOD_RankFeat(Detector):
```
Comment (kkirchheim) [inline examples/benchmarks/interface/midog_openmibood.py:255]: How is this different from pytorch-oods RankFeat?

File: src/pytorch_ood/dataset/img/openmibood.py [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+            zf.extractall(extract_parent)
+
+        if exists(archive_path):
+            os.remove(archive_path)
+
+
+class PhaKIR(OpenMIBOODDataset):
+    """
+    In-distribution laparoscopic surgical phase classification dataset.
+    Restricted access via Zenodo.
```
Comment (kkirchheim) [inline src/pytorch_ood/dataset/img/openmibood.py:268]: I think we should add this as a warning (orange background) specifying that access has to be requested.

File: src/pytorch_ood/dataset/img/openmibood.py [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+            "Automatic download of PhaKIR is not possible because the dataset has restricted access.\n"
+            "Please request access at https://doi.org/10.5281/zenodo.16753918.\n"
+            "Once approved, place the extracted Video_XX folders under your benchmark root."
+        )
+
+
+class EndoVis2018(OpenMIBOODDataset):
+    """
+    EndoVis 2018 robotic surgery dataset (Near-OOD for PhaKIR).
+    Requires Grand Challenge login.
```
Comment (kkirchheim) [inline src/pytorch_ood/dataset/img/openmibood.py:287]: Here, also, add as warning and explain

File: src/pytorch_ood/dataset/img/openmibood.py [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+
+                    add_bbox = [bbox[0] + 100, bbox[1], bbox[2] + 100, bbox[3]]
+                    if _validate_patch(img_dims, annotations, add_bbox):
+                        crop_neg = image.crop(add_bbox)
+                        crop_neg.save(join(img_dir, f"{img_id:03d}_{ann_id}_0.tiff"))
+
+            if exists(img_path):
+                os.remove(img_path)
+
+        shutil.rmtree(tmp_download, ignore_errors=True)
```
Comment (kkirchheim) [inline src/pytorch_ood/dataset/img/openmibood.py:408]: i think we should remove this. there is some danger that users somehow wipe their entire system. it does not seem like this could happen, but better safe then sorry. we should probably just leave the downloaded files there.

File: src/pytorch_ood/dataset/img/openmibood.py [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+            "and download test sequences 1-4 into near/Endovis2018/."
+        )
+
+
+# =====================================================================
+# MIDOG Benchmark Datasets
+# =====================================================================
+
+
+class MIDOG(OpenMIBOODDataset):
```
Comment (kkirchheim) [inline src/pytorch_ood/dataset/img/openmibood.py:308]: For the ID datasets, we should explain how many labels there are.

File: src/pytorch_ood/dataset/img/openmibood.py [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+        tmp_extract = join(self.root, "_tmp_chaos")
+        os.makedirs(tmp_extract, exist_ok=True)
+        log.info(f"Extracting CHAOS -> {tmp_extract}")
+        with zipfile.ZipFile(archive_path, "r") as zf:
+            zf.extractall(tmp_extract)
+
+        os.makedirs(dest_dir, exist_ok=True)
+
+        try:
+            import SimpleITK as sitk
```
Comment (kkirchheim) [inline src/pytorch_ood/dataset/img/openmibood.py:536]: We have to add a not that says that uses have to install this to the docs of this dataset.

File: src/pytorch_ood/dataset/img/openmibood.py [hunk 13]
```diff
@@ -0,0 +1,692 @@
...
+            tf.extractall(os.path.dirname(dest_dir))
+
+        if exists(archive_path):
+            os.remove(archive_path)
+
+
+class OASIS3(OpenMIBOODDataset):
+    """
+    In-distribution brain MRI dataset from OASIS-3.
+    Requires NITRC registration and data use agreement.
```
Comment (kkirchheim) [inline src/pytorch_ood/dataset/img/openmibood.py:602]: Add warning/note regarding access request

File: src/pytorch_ood/detector/fdbd.py [hunk 14]
```diff
@@ -54,14 +54,15 @@ class fDBD(FeaturesDetector):
 
     requires_fit = True
 
-    def __init__(self, encoder: Module, head: Linear) -> None:
+    def __init__(self, encoder: Module, head: Linear, distance_as_normalizer: bool = True) -> None:
```
Comment (kkirchheim) [inline src/pytorch_ood/detector/fdbd.py:57]: This parameter is not documented. what does it do?

File: src/pytorch_ood/detector/gradnorm.py [hunk 15]
```diff
@@ -111,7 +111,10 @@ def predict(self, x: Tensor) -> Tensor:
         x = x.to(device)
 
         if _TORCH_FUNC_AVAILABLE:
-            return self._predict_batched(x)
+            try:
+                return self._predict_batched(x)
+            except Exception:
```
Comment (kkirchheim) [inline src/pytorch_ood/detector/gradnorm.py:116]: What would be the exact exception we are catching here? Also, i think, if we just fall back to sequential processing, we might want to print a warning at least so users are aware something unusual happened.

File: src/pytorch_ood/detector/residual.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
+"""
```
Comment (kkirchheim) [inline src/pytorch_ood/detector/residual.py:1]: So i think this entire class has a huge overlap with the ViM detector. We should check for douplicate code, it can probably be implemented by extending vim and overriding some of its methods.

**Review by kkirchheim — CHANGES_REQUESTED** [review]

File: examples/benchmarks/interface/midog_openmibood.py (hunk 7 above)
Comment (taneeshk12) [reply examples/benchmarks/interface/midog_openmibood.py:255]: The standard PyTorch-OOD RankFeat operates purely on the final output feature map. Our OpenMIBOOD_RankFeat implementation is fundamentally different because it intercepts ResNet's intermediate layer 3 and layer 4 feature maps, removes their rank-1 subspaces independently, and then passes them through the remainder of the network.

File: src/pytorch_ood/detector/gradnorm.py (hunk 15 above)
Comment (taneeshk12) [reply src/pytorch_ood/detector/gradnorm.py:116]: The exception we are catching is a RuntimeError due to CUDA OutOfMemory when the batch size is too large for the vectorized torch.func pass. I have updated the code to explicitly catch RuntimeError and added a log.warning so users are immediately alerted that the detector is falling back to the slower sequential loop

File: examples/benchmarks/interface/midog_openmibood.py (hunk 5 above)
Comment (taneeshk12) [reply examples/benchmarks/interface/midog_openmibood.py:249]: That was the pre-calculated 95th percentile of the intermediate activations on the MIDOG validation set from the paper. However, I agree it shouldn't be hardcoded. I have updated the script to remove the hardcoded value and replaced it with react.fit(train_loader) so that PyTorch-OOD dynamically calculates the exact threshold.

File: src/pytorch_ood/detector/fdbd.py (hunk 14 above)
Comment (taneeshk12) [reply src/pytorch_ood/detector/fdbd.py:57]: The distance_as_normalizer flag controls whether the decision boundary distance is scaled by the instance-wise Euclidean distance from the training mean (mimicking the exact fDBD paper).
